explorer/common/source_location.h

```diff
@@ -13,32 +13,40 @@
 
 namespace Carbon {
 
+// Describes the kind of file that the source location is within.
+enum class FileKind { Main, Prelude, Import, Unknown, Last = Unknown };
+
 class SourceLocation {
  public:
   // Produce a source location that is known to not be used, because it is fed
   // into an operation that creates no AST nodes and whose diagnostics are
   // discarded.
   static auto DiagnosticsIgnored() -> SourceLocation {
-    return SourceLocation("", 0);
+    return SourceLocation("", 0, FileKind::Unknown);
   }
 
   // The filename should be eternal or arena-allocated to eliminate copies.
-  constexpr SourceLocation(const char* filename, int line_num)
-      : filename_(filename), line_num_(line_num) {}
-  SourceLocation(Nonnull<const std::string*> filename, int line_num)
-      : filename_(*filename), line_num_(line_num) {}
+  constexpr SourceLocation(const char* filename, int line_num,
+                           FileKind file_kind)
+      : filename_(filename), line_num_(line_num), file_kind_(file_kind) {}
+  SourceLocation(Nonnull<const std::string*> filename, int line_num,
```

Comment:
```suggestion
  explicit constexpr SourceLocation(const char* filename, int line_num,
                           FileKind file_kind)
      : filename_(filename), line_num_(line_num), file_kind_(file_kind) {}
  explicit SourceLocation(Nonnull<const std::string*> filename, int line_num,
```

Per past explicit constructor style discussions, I think it'll work without affecting callers, and you're modifying them anyways.